### PR TITLE
Make weights normalized

### PR DESCRIPTION
**Pull request by nventuro**
This is a bit of a big change - I didn't find a way to perform this in simpler stages.

All notion of normalization is removed - all tokens are instead always normalized. We drop the normalized sum and compression entirely. This also results in less rounding errors.

There's a few things missing, namely:
 - tests for aum fee collection
 - forbid adding self
 - forbid removing self
 - make tests on new weights more strict (I think we potentially reduce the expected relative error)

### CODE REVIEW

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 1]
```diff
@@ -33,6 +33,8 @@ import "../WeightedMath.sol";
 
 import "./vendor/BasePool.sol";
 
+import "hardhat/console.sol";
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:37]: :construction_worker:

**Review by TomAFrench — APPROVED** [review]
Looks good!

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 2]
```diff
@@ -755,52 +724,88 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
-            ManagedPoolStorageLib.getGradualWeightChangeProgress(_poolState),
-            _denormWeightSum
+        // Deleting the old token is straightforward. The Pool itself doesn't track how many or which tokens it uses
+        // (and relies instead on the Vault for this), so we simply delete the token-specific information. We first read
+        // its weight however, since we'll need it later.
+        uint256 weightChangeProgress = ManagedPoolStorageLib.getGradualWeightChangeProgress(_poolState);
+        uint256 tokenToRemoveWeight = ManagedPoolTokenLib.getTokenWeight(
+            _tokenState[tokenToRemove],
+            weightChangeProgress
         );
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:769]:
```suggestion
        // We've ensured that the most recent weight change is complete.
        uint256 tokenToRemoveWeight = ManagedPoolTokenLib.getTokenWeight(
            _tokenState[tokenToRemove],
            FixedPoint.ONE
        );
```

If you want to optimizoor.

Comment (nventuro) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:769]: Ah interesting. `_getNormalizedWeights` will read the progress again though :)

**Review by EndymionJkb — COMMENTED** [review]
Didn't hit save...

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 3]
```diff
@@ -662,88 +620,97 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
+        _require(tokens.length <= _MAX_TOKENS, Errors.MAX_TOKENS);
+
+        // Once we've updated the state in the Vault, we need to also update our own state. This is a two-step process,
+        // since we need to:
+        //  a) initialize the state of the new token
+        //  b) adjust the weights of all other tokens
+
+        // Initializing the new token is straightforward. The Pool itself doesn't track how many or which tokens it uses
+        // (and relies instead on the Vault for this), so we simply store the new token-specific information.
+        _tokenState[tokenToAdd] = ManagedPoolTokenLib.initializeTokenState(tokenToAdd, tokenToAddNormalizedWeight);
```
Comment (nventuro) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:665]:
```suggestion
        // Note that we don't need to check here that the weight is valid. We'll later call `_startGradualWeightChange`,
        // which will check the entire set of weights for correctness.
        _tokenState[tokenToAdd] = ManagedPoolTokenLib.initializeTokenState(tokenToAdd, tokenToAddNormalizedWeight);
```

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 4]
```diff
@@ -770,51 +739,85 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
-        // revert if we try to deregister a token that is not registered.
-        // We do, however, want to check that the Pool will end up with at least two tokens. This simplifies some
-        // assumptions made elsewhere (e.g. the denormalized weight sum will always be non-zero), and doesn't greatly
-        // restrict the controller.
+        // the case, or if the token is not currently registered.
+        PoolRegistrationLib.deregisterToken(getVault(), getPoolId(), tokenToRemove);
 
         (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
-        _require(tokens.length > 2, Errors.MIN_TOKENS);
+        _require(tokens.length >= 2, Errors.MIN_TOKENS);
```
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:755]: We can't - these are the tokens _after_ deregistration. That could perhaps be made clearer.

**Pull request merged**